nexus/src/context.rs

```diff
@@ -29,6 +29,61 @@ use std::str::FromStr;
 use std::sync::Arc;
 use uuid::Uuid;
 
+/// Indicates the kind of HTTP server.
+#[derive(Clone, Copy)]
+pub enum ServerKind {
+    /// This serves the internal API.
+    Internal,
+    /// This serves the external API over the normal public network.
+    External,
+    /// This serves the external API proxied over the technician port.
+    Techport,
+}
+
+/// The API context for each distinct Dropshot server.
+///
+/// This packages up the main server context, which is shared by all API servers
+/// (e.g., internal, external, and techport). It also includes the
+/// [`ServerKind`], which makes it possible to know which server is handling any
+/// particular request.
+#[derive(Clone)]
+pub struct ApiContext {
+    /// The kind of server.
+    pub kind: ServerKind,
+    /// Shared state available to all endpoint handlers.
+    pub context: Arc<ServerContext>,
+}
+
+impl ApiContext {
+    /// Create a new context with a rack ID and logger. This creates the
+    /// underlying `Nexus` as well.
+    pub async fn new(
+        rack_id: Uuid,
+        log: Logger,
+        config: &NexusConfig,
+    ) -> Result<Self, String> {
+        ServerContext::new(rack_id, log, config)
+            .await
+            .map(|context| Self { kind: ServerKind::Internal, context })
+    }
+
+    /// Clone self for use by the external Dropshot server.
+    pub fn for_external(&self) -> Self {
+        Self { kind: ServerKind::External, context: self.context.clone() }
+    }
+
+    /// Clone self for use by the techport Dropshot server.
+    pub fn for_techport(&self) -> Self {
+        Self { kind: ServerKind::Techport, context: self.context.clone() }
+    }
+}
+
+impl std::borrow::Borrow<ServerContext> for ApiContext {
```

Comment:
We do need borrow here, to satisfy this trait bound:

https://github.com/oxidecomputer/omicron/blob/8c38ad1f07f90b6caca0062a6be5a3aabb5e6e79/nexus/db-queries/src/authn/external/mod.rs#L49-L54

I agree that it seems a bit weird, and I'm not sure I like the trait bound itself. I kept it simply to minimize churn, but I'm happy to reevaluate.

I also considered implementing `Deref`. I think it might make some other changes unnecessary, but there actually didn't seem to be many `ServerContext` methods used themselves. Most of the fields are visible and used directly. I also agree it's a bit magical sometimes, so I think it makes sense to defer it.

Reply:
TIL! I'd definitely defer any reevaluation of that for some followup work (which is probably not necessary at all - implementing `Borrow` is surprising but actually required - that's fine).